sale_coupon_multi_gift/models/sale_order.py

```diff
@@ -0,0 +1,157 @@
+        programs_to_remove = applied_programs - applicable_programs
+        # We're only interested in the Multi Gift programs
+        multi_gift_applied_programs = applied_programs.filtered(
+            lambda x: x.reward_type == "multi_gift"
+        )
+        # These will be the ones to keep
+        valid_lines = self.order_line.filtered(
+            lambda x: x.is_reward_line
+            and x.coupon_program_id in multi_gift_applied_programs
+        )
+        multi_gift_programs_to_remove = programs_to_remove.filtered(
+            lambda x: x.reward_type == "multi_gift"
+        )
+        if multi_gift_programs_to_remove:
+            # Invalidate the generated coupons which we are not eligible anymore
+            self.generated_coupon_ids.filtered(
+                lambda x: x.program_id in multi_gift_programs_to_remove
+            ).write({"state": "expired"})
+            # Detect and remove the proper unvalid program order lines
+            self.order_line.filtered(
+                lambda x: x.is_reward_line
+                and x.coupon_program_id in multi_gift_programs_to_remove
+            ).unlink()
+        # We'll catch the context in the subsequent unlink() method
+        super(
+            SaleOrder, self.with_context(valid_multi_gift_lines=valid_lines.ids)
+        )._remove_invalid_reward_lines()
+
+    def _update_existing_reward_lines(self):
+        """We need to match `multi gift` programs with their discount product"""
+        self.ensure_one()
+        super(
+            SaleOrder, self.with_context(only_reward_lines=True)
+        )._update_existing_reward_lines()
+        applied_programs = self._get_applied_programs_with_rewards_on_current_order()
+        for program in applied_programs.filtered(
+            lambda x: x.reward_type == "multi_gift"
+        ):
+            for reward_line in program.coupon_multi_gift_ids:
+                values = self._get_reward_values_multi_gift_line(reward_line, program)
+                lines = self.order_line.filtered(
+                    lambda line: line.product_id == reward_line.reward_product_id
+                    and line.is_reward_line
+                    and line.coupon_program_id == program
+                )
+                # Remove reward line if price or qty equal to 0
+                if values.get("product_uom_qty") and values.get("price_unit"):
+                    lines.write(values)
+                else:
+                    lines.unlink()
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    def unlink(self):
+        """Avoid unlinking valid multi gift lines since they aren't linked to the
+        discount product of the promotion program"""
+        if not self.env.context.get("valid_multi_gift_lines"):
+            return super().unlink()
+        return super(
+            SaleOrderLine,
+            self.filtered(
+                lambda x: x.id not in self.env.context.get("valid_multi_gift_lines")
```

Comment:
Is this done the same way on core?

Reply:
No, it's explained here: https://github.com/OCA/sale-promotion/blob/a8f67ff597cd8e9d7e58997cda49817b475c9113/sale_coupon_multi_gift/models/sale_order.py#L77-L85